lib/wlib/strings/DynamicString.cpp

```diff
@@ -0,0 +1,482 @@
+/**
+ * @file DynamicString.cpp
+ * @brief DynamicString is a class that provides dynamic strings and functions for dynamic strings
+ *
+ * @author Bob Wei
+ * @date November 25, 2017
+ * @bug No known bugs
+ */
+
+#include "DynamicString.h"
+
+namespace wlp {
+
+    /**
+     * Default constructor creates string with no characters
+     */
+    DynamicString::DynamicString() {
+        m_buffer = malloc<char>();
+        m_buffer[0] = '\0';
+        len = 0;
+    }
+
+    /**
+     * Constructor creates string using character array
+     *
+     * @param str char string
+     */
+    DynamicString::DynamicString(const char *str) {
+        m_buffer = nullptr;
+        auto size = (size_type) strlen(str);
+        m_buffer = malloc<char[]>(static_cast<size_type>(size + 1u));
+        strncpy(m_buffer, str, size);
+        m_buffer[size] = '\0';
+        len = size;
+    }
+
+    /**
+     * Constructor creates string using DynamicString object
+     *
+     * @param str @code DynamicString object
+     */
+    DynamicString::DynamicString(const DynamicString &str) : DynamicString(str.c_str()) {}
+
+    /**
+     * Destructor for DynamicString object
+     */
+    DynamicString::~DynamicString() {

```

Comment:
please explicitly `inline` pretty much every function here except maybe `append(const char *)`

Reply:
If we make it inline the definitions have to be `.h` files so it is fine most new compilers and linkers are fast enough to not notice any difference

Reply:
Inline functions can be defined outside the class i.e. in cpp files.

Reply:
I am fixing this class rn and apparently it does not let me
https://stackoverflow.com/questions/17054367/moving-inline-methods-from-a-header-file-to-a-cpp-files

Reply:
Jk ur right